Constexpr assignment operators in PublicKey and PrivateKey

I have few questions/issues about these copy assignments introduced recently. 

https://github.com/opendnssec/SoftHSMv2/blob/e4c954e96ae5b625ca800b78f9fee5e2cc453423/src/lib/crypto/PrivateKey.h#L71
https://github.com/opendnssec/SoftHSMv2/blob/e4c954e96ae5b625ca800b78f9fee5e2cc453423/src/lib/crypto/PublicKey.h#L64

1) It does not make much sense to define an assignment operator `const` function - you should not assign into a constant variable.
2) The implementation also seems weird. You would not expect an assignment operator to do nothing.
3) According to the standard, assignment operator can be `constexpr` only if all the types involved are `LiteralType`, which the keys are not - they have virtual destructor.

Is there some reason behind it? If not, I suggest to change this to plain:

```c++
PrivateKey& operator= (const PrivateKey&) = default;
```

Or, if the purpose was to silence warnings but disable assignment, then:

```c++
PrivateKey& operator= (const PrivateKey&) = delete;
```

Both should work in all compilers, these are common snippets.

Then the code does quite the opposite :) Assignment is more or less a function call. If you provide an implementation which is `const`, I believe that compiler will just call it, allowing semantically incorrect code to compile. (A compile error would be triggered if you actually attempted to modify `this` inside the function, or e.g. return it as a non-const reference - which is why #535 exists.)

The original warning is due to absence of any copy assignment operator - if you explicitly say that it is deleted, it will dismiss the warning without allowing copy assignment.

At the very least, we have to remove the `constexpr`, because it triggers a compile error in e.g. GCC 6.3:

```
./../PublicKey.h:64:29: error: enclosing class of constexpr non-static member function 'const PublicKey& PublicKey::operator=(const PublicKey&) const' is not a literal type
  constexpr const PublicKey& operator=(const PublicKey&) const { return *this; }
                             ^~~~~~~~
./../PublicKey.h:40:7: note: 'PublicKey' is not literal because:
 class PublicKey : public Serialisable
       ^~~~~~~~~
./../PublicKey.h:40:7: note:   'PublicKey' has a non-trivial descructor
```

... which is because of the limitation described in 3).

--- a/src/lib/SoftHSM.cpp
+++ b/src/lib/SoftHSM.cpp
@@ -1080,6 +1080,7 @@ CK_RV SoftHSM::C_GetMechanismInfo(CK_SLOT_ID slotID, CK_MECHANISM_TYPE type, CK_
 #endif
 		case CKM_DES3_CBC:
 			pInfo->flags = CKF_WRAP;
+			// falls through
 		case CKM_DES3_ECB:
 		case CKM_DES3_CBC_PAD:
 			// Key size is not in use
@@ -1100,6 +1101,7 @@ CK_RV SoftHSM::C_GetMechanismInfo(CK_SLOT_ID slotID, CK_MECHANISM_TYPE type, CK_
 			break;
 		case CKM_AES_CBC:
 			pInfo->flags = CKF_WRAP;
+			// falls through
 		case CKM_AES_ECB:
 		case CKM_AES_CBC_PAD:
 		case CKM_AES_CTR:

--- a/src/lib/crypto/AsymmetricKeyPair.h
+++ b/src/lib/crypto/AsymmetricKeyPair.h
@@ -42,14 +42,6 @@
 class AsymmetricKeyPair : public Serialisable
 {
 public:
-	// Base constructors
-	AsymmetricKeyPair() { }
-
-	AsymmetricKeyPair(const AsymmetricKeyPair& /*in*/) { }
-
-	// Destructor
-	virtual ~AsymmetricKeyPair() { }
-
 	// Return the public key
 	virtual PublicKey* getPublicKey() = 0;
 	virtual const PublicKey* getConstPublicKey() const = 0;

--- a/src/lib/crypto/AsymmetricParameters.h
+++ b/src/lib/crypto/AsymmetricParameters.h
@@ -40,19 +40,8 @@
 class AsymmetricParameters : public Serialisable
 {
 public:
-	// Base constructors
-	AsymmetricParameters() { }
-
-	AsymmetricParameters(const AsymmetricParameters& /*in*/) { }
-
-	// Destructor
-	virtual ~AsymmetricParameters() { }
-
 	// Check if it is of the given type
 	virtual bool areOfType(const char* type) = 0;
-
-	// Serialisation
-	virtual ByteString serialise() const = 0;
 };
 
 #endif // !_SOFTHSM_V2_ASYMMETRICPARAMETERS_H

--- a/src/lib/crypto/PrivateKey.h
+++ b/src/lib/crypto/PrivateKey.h
@@ -41,14 +41,6 @@
 class PrivateKey : public Serialisable
 {
 public:
-	// Base constructors
-	PrivateKey() { }
-
-	PrivateKey(const PrivateKey& in);
-
-	// Destructor
-	virtual ~PrivateKey() { }
-
 	// Check if the private key is of the given type
 	virtual bool isOfType(const char* inType) = 0;
 
@@ -58,17 +50,11 @@ public:
 	// Get the output length
 	virtual unsigned long getOutputLength() const = 0;
 
-	// Serialisation
-	virtual ByteString serialise() const = 0;
-
 	// Encode into PKCS#8 DER
 	virtual ByteString PKCS8Encode() = 0;
 
 	// Decode from PKCS#8 BER
 	virtual bool PKCS8Decode(const ByteString& ber) = 0;
-
-	// Assignment
-	constexpr const PrivateKey& operator=(const PrivateKey&) const { return *this; }
 };
 
 #endif // !_SOFTHSM_V2_PRIVATEKEY_H

--- a/src/lib/crypto/PublicKey.h
+++ b/src/lib/crypto/PublicKey.h
@@ -40,11 +40,6 @@
 class PublicKey : public Serialisable
 {
 public:
-	// Base constructors
-	PublicKey() { }
-
-	PublicKey(const PublicKey& /*in*/) { }
-
 	// Destructor
 	virtual ~PublicKey() { }
 
@@ -56,12 +51,6 @@ public:
 
 	// Get the output length
 	virtual unsigned long getOutputLength() const = 0;
-
-	// Serialisation
-	virtual ByteString serialise() const = 0;
-
-	// Assignment
-	constexpr const PublicKey& operator=(const PublicKey&) const { return *this; }
 };
 
 #endif // !_SOFTHSM_V2_PUBLICKEY_H

--- a/src/lib/crypto/SymmetricKey.cpp
+++ b/src/lib/crypto/SymmetricKey.cpp
@@ -42,12 +42,6 @@ SymmetricKey::SymmetricKey(size_t inBitLen /* = 0 */)
 	bitLen = inBitLen;
 }
 
-SymmetricKey::SymmetricKey(const SymmetricKey& in)
-{
-	keyData = in.keyData;
-	bitLen = in.bitLen;
-}
-
 // Set the key
 bool SymmetricKey::setKeyBits(const ByteString& keybits)
 {

--- a/src/lib/crypto/SymmetricKey.h
+++ b/src/lib/crypto/SymmetricKey.h
@@ -40,13 +40,8 @@
 class SymmetricKey : public Serialisable
 {
 public:
-	// Base constructors
-	SymmetricKey(size_t inBitLen = 0);
-
-	SymmetricKey(const SymmetricKey& in);
-
-	// Destructor
-	virtual ~SymmetricKey() { }
+	// Constructor
+	explicit SymmetricKey(size_t inBitLen = 0);
 
 	// Set the key
 	virtual bool setKeyBits(const ByteString& keybits);

--- a/src/lib/object_store/DB.cpp
+++ b/src/lib/object_store/DB.cpp
@@ -487,7 +487,7 @@ time_t DB::Result::getDatetime(unsigned int fieldidx)
 	switch (valuelen) {
 		case 19:	// 2011-12-31 23:59:59
 			bdateonly = false;
-			// falls through to next case
+			// falls through
 		case 10:	// 2011-12-31
 			years = strtoul(p,&pnext,10);
 			gm_tm.tm_year = ((int)years)-1900; /* years since 1900 */
@@ -500,7 +500,7 @@ time_t DB::Result::getDatetime(unsigned int fieldidx)
 			p = pnext+1;
 			if (bdateonly)
 				break;
-			// falls through to next case
+			// falls through
 		case 8:		// 23:59:59
 			hours = strtoul(p,&pnext,10);
 			gm_tm.tm_hour = (int)hours; /* hours since midnight [0-23] */

--- a/src/lib/object_store/UUID.h
+++ b/src/lib/object_store/UUID.h
@@ -42,7 +42,7 @@ namespace UUID
 {
 	// Generate a new UUID string
 	std::string newUUID();
-};
+}
 
 #endif // !_SOFTHSM_V2_UUID_H
 

